src/main/java/com/coveo/spillway/storage/utils/CacheSynchronisation.java

```diff
@@ -0,0 +1,96 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.coveo.spillway.storage.utils;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.TimerTask;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.coveo.spillway.limit.LimitKey;
+import com.coveo.spillway.storage.AsyncBatchLimitUsageStorage;
+import com.coveo.spillway.storage.InMemoryStorage;
+import com.coveo.spillway.storage.LimitUsageStorage;
+
+/**
+ * Synchronization {@link TimerTask} that is launched periodically
+ * by the {@link AsyncBatchLimitUsageStorage}.
+ *
+ * @author Emile Fugulin
+ * @since 1.0.0
+ */
+public class CacheSynchronisation extends TimerTask {
+  private static final Logger logger = LoggerFactory.getLogger(CacheSynchronisation.class);
+
+  private InMemoryStorage cache;
+  private LimitUsageStorage storage;
+
+  public CacheSynchronisation(InMemoryStorage cache, LimitUsageStorage storage) {
+    this.cache = cache;
+    this.storage = storage;
+  }
+
+  @Override
+  public void run() {
+    cache.applyOnEach(
+        (instantEntry) -> {
+          try {
+            Instant expiration = instantEntry.getKey();
+
+            instantEntry
+                .getValue()
+                .entrySet()
+                .forEach(
+                    valueEntry -> {
+                      int cost = valueEntry.getValue().getDelta();
+
+                      LimitKey limitKey = valueEntry.getKey();
+
+                      Duration duration =
+                          Duration.ofMillis(
+                              expiration.toEpochMilli() - limitKey.getBucket().toEpochMilli());
+
+                      AddAndGetRequest request =
+                          new AddAndGetRequest.Builder()
+                              .withResource(limitKey.getResource())
+                              .withLimitName(limitKey.getLimitName())
+                              .withProperty(limitKey.getProperty())
+                              .withExpiration(duration)
+                              .withEventTimestamp(limitKey.getBucket())
+                              .withCost(cost)
+                              .build();
+
+                      Pair<LimitKey, Integer> reponse = storage.addAndGet(request);
```

Comment:
I agree that pipelining the queries would be better, but I don't really know how to do that while iterating over the map. We have to remember that it is less critical since its an async job and we don't have that many entries.